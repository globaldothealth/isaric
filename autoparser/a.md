# autoparser configuration schema

*Schema for autoparser configuration*

## Properties

- **`name`** *(string)*: Name of the configuration.
- **`description`** *(string)*: Description of the configuration.
- **`choice_delimiter`** *(string)*: Delimiter used to separate integer -> value mappings. Used by parse_choices() to generate values mapping.
- **`choice_delimiter_map`** *(string)*: Delimiter used to separate integer from value. Used by parse_choices() to generate values mapping.
- **`categorical_types`** *(array)*: List of categorical types according to data dictionary. Categorical field types are used by single_field_mapping() to infer whether values or references to values mapping should be set.
  - **Items** *(string)*
- **`text_types`** *(array)*: List of non-categorical (textual) types according to data dictionary. Text field types are used as a hint to single_field_mapping() to _not_ set a values mapping.
  - **Items** *(string)*
- **`lang`** *(object)*: Language specific settings.
  - **`is_missing`** *(array)*: Denotes missing values.
    - **Items** *(string)*
  - **`is_true`** *(array)*: Denotes true or yes values.
    - **Items** *(string)*
  - **`is_false`** *(array)*: Denotes false or no values.
    - **Items** *(string)*
  - **`stopwords`** *(array)*: Words to ignore when matching or when generating TF-IDF matrix.
    - **Items** *(string)*
- **`schemas`** *(object)*: Schema mappings.
  - **`subject`** *(string)*: Schema for the subject mapping.
  - **`visit`** *(string)*: Schema for the visit mapping.
  - **`observation`** *(string)*: Schema for the observation mapping.
- **`scores`** *(object)*: Scores used by csv_mapping to order mappings.
  - **`*`** *(integer)*
- **`column_mappings`** *(object)*: Mappings of intermediate CSV to source data dictionary fields.
  - **`schema_field`** *(string)*: Field in the ISARIC schema that this mapping corresponds to.
  - **`field`** *(string)*: Field in the source data file that corresponds to the ISARIC schema field.
  - **`category`** *(string)*: Field category according to source.
  - **`description`** *(string)*: Field description.
  - **`choices`** *(string)*: Delimited field -> value mappings.
  - **`note`** *(string)*: Field note.
  - **`valid_type`** *(string)*: Field validation type according to source.
  - **`valid_min`** *(string)*: Minimum value for field if applicable.
  - **`valid_max`** *(string)*: Whether field is identifier.
  - **`is_identifier`** *(string)*: Whether field is identifier.
  - **`condition`** *(string)*: Field condition, when field is shown.
  - **`is_required`** *(string)*: Whether field is required in the source form.
- **`observation_type_mapping`** *(object)*: Mapping from source types to observation fields.
  - **`*`**: Must be one of: `['is_present', 'value', 'text']`.
