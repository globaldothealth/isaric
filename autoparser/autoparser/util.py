"""
Common utility functions for autoparser
"""

DEFAULT_CONFIG = "config/redcap-en.toml"


def maybe(x, func, default=None):
    return func(x) if x is not None else default
