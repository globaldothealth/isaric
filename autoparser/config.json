{
  "stopwords": [
    "a",
    "an",
    "the",
    "if",
    "for",
    "has",
    "at",
    "of",
    "and",
    "or",
    "then",
    "treatment",
    "agent"
  ],
  "schemas": {
    "subject": "../schemas/dev/subject.schema.json",
    "visit": "../schemas/dev/visit.schema.json",
    "observation": "../schemas/dev/observation.schema.json"
  },
  "column_mappings": {
    "Variable / Field Name": "source_field",
    "Form Name": "category",
    "Section Header": "section_header",
    "Field Type": "source_field_type",
    "Field Label": "source_field_description",
    "Choices, Calculations, OR Slider Labels": "choices",
    "Field Note": "source_field_note",
    "Text Validation Type OR Show Slider Number": "source_validation_type",
    "Text Validation Min": "source_valid_min",
    "Text Validation Max": "source_valid_max",
    "Identifier?": "source_field_is_identifier",
    "Branching Logic (Show field only if...)": "source_condition",
    "Required Field?": "source_field_is_required"
  },
  "scores": {
    "type-match": 3,
    "date-mismatch": -3,
    "type-mismatch": -1,
    "token-contained": 1,
    "is-followup": -3
  },
  "use_fields": {
    "observation": "properties.name.enum"
  }
}