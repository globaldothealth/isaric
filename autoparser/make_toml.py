"""
Generate TOML parser from intermediate CSV file
"""
import re
import json
import argparse
import operator
from pathlib import Path

import tomli_w
import pandas as pd

MISSING = [
    "unknown",
    "n/a",
    "n/k",
    "na",
    "nk",
    "not applicable",
    "prefer not to say",
    "not specified",
    "not answered",
]
TRUE = ["y", "yes", "t", "true"]
FALSE = ["f", "false", "no", "n"]


def maybe(x, func, default=None):
    return func(x) if x is not None else default


def read_data(path: Path) -> dict:
    if path.suffix == ".json":
        with path.open() as fp:
            return json.load(fp)
    elif path.suffix == ".toml":
        with path.open("rb") as fp:
            return tomli.load(fp)


def adtl_header(
    name: str, description: str, schema_version: str = "dev", definitions={}
):
    return {
        "adtl": {
            "name": name,
            "description": description,
            "tables": {
                "study": {"kind": "constant"},
                "subject": {
                    "kind": "groupBy",
                    "groupBy": "subject_id",
                    "aggregation": "lastNotNull",
                    "schema": f"../../schemas/{schema_version}/subject.schema.json",
                },
                "visit": {
                    "kind": "groupBy",
                    "groupBy": "subject_id",
                    "aggregation": "lastNotNull",
                    "schema": f"../../schemas/{schema_version}/visit.schema.json",
                },
                "observation": {
                    "kind": "oneToMany",
                    "schema": f"../../schemas/{schema_version}/observation.schema.json",
                    "common": {"visit_id": {"field": "subjid", "sensitive": True}},
                },
            },
            **{"defs": definitions},
        }
    }


def parse_choices(s: str, delimiter: str, delimiter_map: str) -> dict[str]:
    lower_string = lambda s: s.strip().lower()
    if not isinstance(s, str):
        return None
    choices = dict(
        tuple(map(lower_string, x.split(delimiter_map)[:2])) for x in s.split(delimiter)
    )
    # drop n/a, n/k, unknowns
    choices = {k: v for k, v in choices.items() if v not in MISSING}
    for k, v in choices.items():
        if v in TRUE:
            choices[k] = True
        if v in FALSE:
            choices[k] = False
    return choices


def make_toml_table(
    config: dict[str],
    mappings: pd.DataFrame,
    table: str,
    references: dict[str],
    choice_delimiter="|",
    choice_delimiter_map=",",
) -> dict[str]:
    "Make TOML table (not observation, which is handled separately)"
    assert table in set(config["schemas"]) - {"observation"}
    schema = read_data(config["schemas"][table])["properties"]
    mappings = mappings[mappings.table == table]
    if mappings.empty:
        return {}
    outmap = {}

    def single_field_mapping(
        match: pd.core.frame.pandas, field_type: str, field_enum: list[str] = []
    ) -> dict[str]:
        out = {
            "field": match.source_field,
            "description": match.source_field_description,
        }
        if field_type in ["boolean", "enum"] and (
            choices := parse_choices(
                match.choices, choice_delimiter, choice_delimiter_map
            )
        ):
            if (choice_key := json.dumps(choices, sort_keys=True)) in references:
                out["ref"] = references[choice_key]
            else:
                out["values"] = choices

        if field_type == "enum" and field_enum and out.get("values"):
            mapped_enum = map_enum(list(out["values"].values()), field_enum)
            source_values = out["values"].copy()
            out["values"] = {
                k: mapped_enum[v] for k, v in out["values"].items() if v in mapped_enum
            }
            print(
                f"{match.source_field}\n"
                + "\n".join(
                    f"    {source} -> {target}"
                    for source, target in zip(
                        source_values.values(), out["values"].values()
                    )
                )
            )

        return out

    for field, field_matches in mappings.groupby("field"):
        field_type = schema[field].get("type")
        field_enum = schema[field].get("enum", [])
        if "enum" in schema[field]:
            field_type = "enum"
        if len(field_matches) == 1:  # single field
            outmap[field] = single_field_mapping(
                field_matches.iloc[0], field_type, field_enum
            )

        else:  # combinedType
            outmap[field] = {
                "combinedType": {"array": "list", "boolean": "any"}.get(
                    schema[field].get("type"), "firstNonNull"
                ),
                "fields": [
                    single_field_mapping(match, field_type, field_enum)
                    for match in field_matches.itertuples()
                ],
            }

    return {table: outmap}


def make_toml_observation(
    config: dict[str],
    mappings: pd.DataFrame,
) -> list[dict[str]]:
    mappings = mappings[mappings.table == "observation"]
    observation = []


def make_toml(
    config: dict[str],
    mappings: pd.DataFrame,
    name: str,
    tables: list[str] = ["subject", "visit"],
    description: str = None,
):
    references, definitions = common_mappings(
        mappings, delimiter="|", delimiter_map=","
    )
    data = adtl_header(name, description or name, definitions=definitions)
    for table in tables:
        if table != "observation":
            data.update(make_toml_table(config, mappings, table, references))
        else:
            data.update(make_toml_observation(config, mappings))
    return data


def map_enum(keys: list[str], sources: list[str]) -> dict[str, str]:
    def score(key, source):
        key = re.split("-|_| ", str(key).lower())
        source = re.split("-|_| ", source.lower())
        return sum(
            sum([[s.startswith(w) or s.endswith(w) for s in source] for w in key], [])
        )

    def max_score(key, sources):
        try:
            return max(
                (
                    (source, score(key, source))
                    for source in sources
                    if score(key, source) > 0
                ),
                key=operator.itemgetter(1),
            )[0]
        except ValueError:
            return None

    mapped_enum = {key: max_score(key, sources) for key in keys}
    return {k: v for k, v in mapped_enum.items() if v is not None}


def common_mappings(
    mappings: pd.DataFrame, delimiter: str, delimiter_map: str, top: int = 3
) -> tuple[dict[str], dict[str]]:
    mappings.choices.value_counts()[:3]
    references = {}
    definitions = {}

    def _parse_choices(s):
        c = parse_choices(s, delimiter, delimiter_map)
        return json.dumps(c, sort_keys=True) if c else None

    # use value_counts() on parsed_choices instead of choices to
    # normalise various flavours of Y/N/NK or Y/N/NA
    mappings["parsed_choices"] = mappings.choices.map(_parse_choices)
    top_mappings = mappings.parsed_choices.value_counts()[:top].index
    boolean_mapping_name = "Y/N/NK"

    # only add one boolean map for simplicity
    boolean_map_found = False
    for mapping in top_mappings:
        if boolean_map_found and "true" in mapping:
            continue
        if "true" in mapping:
            references[mapping] = "Y/N/NK"
            definitions["Y/N/NK"] = {"values": json.loads(mapping)}
            boolean_map_found = True
            continue
        c = json.loads(mapping)
        name = "/".join(map(str, c.values()))
        references[mapping] = name
        definitions[name] = {"values": c}
    return references, definitions


def write_toml(data: dict[str], output: str):
    with open(output, "wb") as fp:
        tomli_w.dump(data, fp)


def main():
    parser = argparse.ArgumentParser(
        description="Make TOML from intermediate CSV file created by csv_mapping.py"
    )
    parser.add_argument(
        "mappings", help="Mapping file(s) to create TOML from", nargs="+", type=str
    )
    parser.add_argument(
        "-n", "--name", help="Name of the parser (default=isaric)", default="isaric"
    )
    parser.add_argument(
        "-c",
        "--config",
        help="Configuration file to use (default=config.toml)",
        type=Path,
    )

    args = parser.parse_args()
    mappings = pd.concat((pd.read_csv(f) for f in args.mappings), ignore_index=True)

    with open(f"{args.name}.toml", "wb") as fp:
        tomli_w.dump(
            make_toml(
                read_data(
                    maybe(
                        args.config,
                        Path,
                        default=Path(__file__).with_name("config.toml"),
                    )
                ),
                mappings,
                args.name,
            ),
            fp,
        )


if __name__ == "__main__":
    main()
