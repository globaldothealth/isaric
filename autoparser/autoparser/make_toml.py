"""
Generate TOML parser from intermediate CSV file
"""
import os
import re
import json
import argparse
import functools
import logging
import operator
from pathlib import Path
from typing import Dict, List, Any, Tuple

import pandas as pd

from .util import maybe, parse_choices, read_data, DEFAULT_CONFIG
from .toml_writer import dump


def adtl_header(
    name: str, description: str, schema_version: str = "dev", definitions={}
):
    return {
        "adtl": {
            "name": name,
            "description": description,
            "tables": {
                "subject": {
                    "kind": "groupBy",
                    "groupBy": "subject_id",
                    "aggregation": "lastNotNull",
                    "schema": f"../../schemas/{schema_version}/subject.schema.json",
                },
                "visit": {
                    "kind": "groupBy",
                    "groupBy": "subject_id",
                    "aggregation": "lastNotNull",
                    "schema": f"../../schemas/{schema_version}/visit.schema.json",
                },
                "observation": {
                    "kind": "oneToMany",
                    "schema": f"../../schemas/{schema_version}/observation.schema.json",
                    "common": {"visit_id": {"field": "subjid", "sensitive": True}},
                },
            },
            **{"defs": definitions},
        }
    }


@functools.lru_cache
def get_schema(schema_path: Path) -> Dict[str, Any]:
    return read_data(schema_path)["properties"]


def get_type_enum(config: Dict[str, Any], table: str, field: str):
    "Returns field types and enums from a field in a table"

    schema = get_schema(
        schema_path := Path(config["schema-path"]) / config["schemas"][table]
    )
    if field not in schema:
        raise ValueError(
            f"get_type_enum(): {field!r} field not in {table!r} schema, from {schema_path}"
        )
    if "enum" in schema[field]:
        return "enum", schema[field]["enum"]
    else:
        return schema[field].get("type", "string"), []


def single_field_mapping(
    config: Dict[str, Any],
    match: pd.core.frame.pandas,
    references: Dict[str, Any],
    use_type: str = None,
    add_auto_condition=False,
) -> Dict[str, Any]:
    try:
        choices, nulls = parse_choices(config, match.choices)
    except ValueError as e:
        logging.warning(e)
        choices, nulls = None, []
    out = {"field": match.field, "description": match.description}
    if not use_type:  # auto detect type from schema field type
        schema_field_type, schema_field_enum = get_type_enum(
            config, match.table, match.schema_field
        )
    else:
        schema_field_type, schema_field_enum = use_type, []
    if match.type in config["categorical_types"] and choices:
        if (choice_key := json.dumps(choices, sort_keys=True)) in references:
            out["ref"] = references[choice_key]
        else:
            out["values"] = choices

    if (
        schema_field_type == "enum"
        and schema_field_enum
        and (out.get("values") or out.get("ref"))
    ):
        if out.get("ref"):
            # re-expand references to enable map_enum() to work
            backref = [k for k in references if references[k] == out["ref"]][0]
            out["values"] = json.loads(backref)
            del out["ref"]
        mapped_enum = map_enum(list(out["values"].values()), schema_field_enum)
        out["values"] = {
            k: mapped_enum[v] for k, v in out["values"].items() if v in mapped_enum
        }
        if mapped_enum:
            print(
                f"\n* {match.schema_field!r} from {match.field!r}, one of {schema_field_enum!r}"
            )
            for s, t in mapped_enum.items():
                print(f"  - {s} â†’ {t}")

    # Only add auto-generated condition if flag enabled and we have not already
    # added a condition, usually from a condition column in the data dictionary
    if add_auto_condition and "if" not in out:
        if "values" not in out or not nulls:  # text field
            out["if"] = {match.field: {"!=": ""}}
        elif len(nulls) > 1:
            out["if"] = {"all": [{match.field: {"!=": n}} for n in nulls]}
        else:
            out["if"] = {match.field: {"!=": nulls[0]}}
        if out.get("ref") == "Y/N/NK":  # most common case
            out["if"] = {match.field: {"!=": 3}}

    return out


def make_toml_table(
    config: Dict[str, Any],
    mappings: pd.DataFrame,
    table: str,
    references: Dict[str, Any],
) -> Dict[str, Any]:
    "Make TOML table (observation handling is separate)"
    if table == "observation":
        return _make_toml_observation(config, mappings, table, references)

    mappings = mappings[mappings.table == table]
    if mappings.empty:
        return {}
    outmap = {}

    for field, field_matches in mappings.groupby("schema_field"):
        field_type, field_enum = get_type_enum(config, table, field)
        if len(field_matches) == 1:  # single field
            outmap[field] = single_field_mapping(
                config,
                field_matches.iloc[0],
                references,
            )

        else:  # combinedType
            if isinstance(field_type, list):  # field can be of multiple types
                if "array" in field_type:
                    field_type = "array"
                elif "boolean" in field_type:
                    field_type = "boolean"
                else:
                    field_type = field_type[0]
            outmap[field] = {
                "combinedType": {"array": "list", "boolean": "any"}.get(
                    field_type, "firstNonNull"
                ),
                "fields": [
                    single_field_mapping(
                        config,
                        match,
                        references,
                    )
                    for match in field_matches.itertuples()
                ],
            }

    return {table: outmap}


def _make_toml_observation(
    config: Dict[str, Any],
    mappings: pd.DataFrame,
    table: str,
    references: Dict[str, Any],
) -> Dict[str, Any]:
    assert table == "observation"
    mappings = mappings[mappings.table == "observation"]
    observations = []

    for mapping in mappings.itertuples():
        if "category" in mappings.columns:
            phase = map_enum(
                [mapping.category.replace("treatment", "study")],
                ["admission", "study", "followup"],
            )
            phase = phase.get(mapping.category, "study")
        else:
            phase = "study"  # no category information found, default to study
        obs_type = config["observation_type_mapping"].get(mapping.type, "value")
        field_mapping = single_field_mapping(
            config, mapping, references, use_type=mapping.type, add_auto_condition=True
        )

        # Move if to observation block instead of in the mapping
        condition = field_mapping["if"]
        del field_mapping["if"]
        observations.append(
            {
                "name": mapping.schema_field,
                "phase": phase,
                "date": {
                    "ref": {
                        "admission": "admissionDateHierarchy",
                        "study": "dailyDateHierarchy",
                        "followup": "followupDateHierarchy",
                    }.get(phase)
                },
                "if": condition,
                obs_type: field_mapping,
            }
        )
    return {"observation": observations}


def make_toml(
    config: Dict[str, Any],
    mappings: pd.DataFrame,
    name: str,
    tables: List[str] = ["subject", "visit", "observation"],
    description: str = None,
):
    references, definitions = common_mappings(config, mappings)
    data = adtl_header(name, description or name, definitions=definitions)
    for table in tables:
        data.update(make_toml_table(config, mappings, table, references))
    return data


def map_enum(keys: List[str], sources: List[str]) -> Dict[str, str]:
    def score(key, source):
        key = re.split("-|_| ", str(key).lower())
        source = re.split("-|_| ", source.lower())
        return sum(
            sum(
                [
                    [int(s.startswith(w) or s.endswith(w)) * len(w) for s in source]
                    for w in key
                ],
                [],
            )
        )

    def max_score(key, sources):
        try:
            return max(
                (
                    (source, score(key, source))
                    for source in sources
                    if score(key, source) > 0
                ),
                key=operator.itemgetter(1),
            )[0]
        except ValueError:
            return None

    mapped_enum = {key: max_score(key, sources) for key in keys}
    return {k: v for k, v in mapped_enum.items() if v is not None}


def common_mappings(
    config: Dict[str, Any], mappings: pd.DataFrame, top: int = 3
) -> Tuple[Dict[str, Any], Dict[str, Any]]:
    mappings.choices.value_counts()[:3]
    references = {}
    definitions = {}

    def _parse_choices(s):
        c, _ = parse_choices(config, s)
        return json.dumps(c, sort_keys=True) if c else None

    # use value_counts() on parsed_choices instead of choices to
    # normalise various flavours of Y/N/NK or Y/N/NA
    mappings["parsed_choices"] = mappings.choices.map(_parse_choices)
    top_mappings = mappings.parsed_choices.value_counts()[:top].index

    # only add one boolean map for simplicity
    boolean_map_found = False
    for mapping in top_mappings:
        if boolean_map_found and "true" in mapping:
            continue
        if "true" in mapping:
            references[mapping] = "Y/N/NK"
            definitions["Y/N/NK"] = {"values": json.loads(mapping)}
            boolean_map_found = True
            continue
        c = json.loads(mapping)
        name = "/".join(map(str, c.values()))
        references[mapping] = name
        definitions[name] = {"values": c}
    return references, definitions


def write_toml(data: Dict[str, Any], output: str):
    with open(output, "wb") as fp:
        dump(data, fp)


def main():
    parser = argparse.ArgumentParser(
        description="Make TOML from intermediate CSV file created by csv_mapping.py"
    )
    parser.add_argument(
        "mappings", help="Mapping file(s) to create TOML from", nargs="+", type=str
    )
    parser.add_argument(
        "-n", "--name", help="Name of the parser (default=isaric)", default="isaric"
    )
    parser.add_argument(
        "-c",
        "--config",
        help=f"Configuration file to use (default={DEFAULT_CONFIG})",
        type=Path,
    )
    parser.add_argument("--schema-path", help="Path where ISARIC schemas are located")

    args = parser.parse_args()
    mappings = pd.concat((pd.read_csv(f) for f in args.mappings), ignore_index=True)
    config = read_data(
        maybe(args.config, Path, default=Path(__file__).parent / DEFAULT_CONFIG)
    )
    config["schema-path"] = (
        maybe(args.schema_path, Path)
        or maybe(os.getenv("ISARIC_SCHEMA_PATH"), Path)
        or Path.cwd()
    )

    write_toml(make_toml(config, mappings, args.name), f"{args.name}.toml")


if __name__ == "__main__":
    main()
