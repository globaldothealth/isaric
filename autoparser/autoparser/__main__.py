import sys

from .csv_mapping import main as csv_mapping_main
from .make_toml import main as make_toml_main


def main():
    if len(sys.argv) < 2:
        print(
            """autoparser: specify subcommand to run

Available subcommands:
  create-parser - Generate TOML parser from CSV mapping
  create-mapping - Create initial CSV mapping from data dictionary
"""
        )
        sys.exit(1)
    subcommand = sys.argv[1]
    if subcommand not in ["create-parser", "create-mapping"]:
        print("autoparser: unrecognised subcommand", subcommand)
        sys.exit(1)
    sys.argv = sys.argv[1:]
    if subcommand == "create-parser":
        make_toml_main()
    elif subcommand == "create-mapping":
        csv_mapping_main()
    else:
        pass


if __name__ == "__main__":
    main()
