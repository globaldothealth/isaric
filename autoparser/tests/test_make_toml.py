"""
Tests for autoparser-toml (make_toml.py)
"""

import pytest
import pandas as pd
from pathlib import Path

import autoparser.make_toml as make_toml
import autoparser.toml_writer as toml_writer

CONFIG = {
    "schema-path": "..",
    "schemas": {
        "subject": "schemas/dev/subject.schema.json",
        "visit": "schemas/dev/visit.schema.json",
        "observation": "schemas/dev/observation.schema.json",
    },
    "categorical_types": ["boolean", "enum", "radio", "dropdown"],
    "text_types": ["text"],
    "choice_delimiter": "|",
    "choice_delimiter_map": ",",
    "lang": {
        "is_missing": [
            "unknown",
            "n/a",
            "n/k",
            "na",
            "nk",
            "not applicable",
            "prefer not to say",
            "not specified",
            "not answered",
        ],
        "is_true": ["y", "yes", "t", "true"],
        "is_false": ["f", "false", "no", "n"],
    },
    "observation_type_mapping": {
        "categorical": "text",
        "boolean": "is_present",
        "yesno": "is_present",
        "radio": "is_present",
        "decimal": "value",
        "integer": "value",
    },
}


@pytest.mark.parametrize(
    "source,expected",
    [
        ("1, yes | 2, no | 3, unknown", ({"1": True, "2": False}, ["3"])),
        ("1, yes | 0, no | 3, unknown", ({"1": True, "0": False}, ["3"])),
        ("yes | no | unknown", (None, [])),
        (
            "kg, weight in kilograms | lbs, weight in pounds",
            ({"kg": "weight in kilograms", "lbs": "weight in pounds"}, []),
        ),
    ],
)
def test_parse_choices(source, expected):
    assert make_toml.parse_choices(CONFIG, source) == expected


def test_map_enum():
    assert make_toml.map_enum(
        ["followup_signs", "info_at_admission2", "followup_symptoms", "during_study5"],
        ["admission", "study", "followup"],
    ) == {"followup_signs": "followup", "followup_symptoms": "followup"}


@pytest.mark.parametrize(
    "source,expected",
    [
        [
            ("observation", "phase"),
            ("enum", ["pre-admission", "admission", "study", "followup"]),
        ]
    ],
)
def test_get_type_enum(source, expected):
    assert make_toml.get_type_enum(CONFIG, *source) == expected


@pytest.mark.parametrize(
    "mapping,expected",
    [
        (
            dict(
                schema_field="sex_at_birth",
                field="sex",
                table="subject",
                category="demographics",
                description="Sex at birth",
                type="radio",
                choices="1, Male | 2, Female | 3, Non-binary | 4, Prefer not to say",
            ),
            dict(
                field="sex",
                description="Sex at birth",
                values={"1": "male", "2": "female", "3": "non_binary"},
            ),
        ),
        (
            dict(
                schema_field="age",
                field="age_estimateyears",
                table="subject",
                category="demographics",
                description="Age/Estimated age",
                type="text",
                choices=None,
            ),
            dict(
                field="age_estimateyears",
                description="Age/Estimated age",
            ),
        ),
    ],
)
def test_single_field_mapping(mapping, expected):
    assert make_toml.single_field_mapping(CONFIG, pd.Series(mapping), {}) == expected


def test_single_field_mapping_with_refs():
    references = {'{"1": true, "2": false}': "Y/N/NK"}
    match = dict(
        schema_field="has_chronic_kidney_disease",
        table="subject",
        field="renal_mhyn",
        type="radio",
        category="comorbidities",
        description="Chronic kidney disease",
        choices="1, YES | 2, NO | 3, Unknown",
        is_required=True,
    )
    assert make_toml.single_field_mapping(CONFIG, pd.Series(match), references) == dict(
        field="renal_mhyn",
        description="Chronic kidney disease",
        ref="Y/N/NK",
    )


def test_make_toml(snapshot):  # integration test
    mappings = pd.concat(
        (
            pd.read_csv(Path(__file__).with_name(f))
            for f in [
                "isaric-subject.csv",
                "isaric-visit.csv",
                "isaric-observation.csv",
            ]
        ),
        ignore_index=True,
    )
    toml_data = make_toml.make_toml(
        CONFIG, mappings, name="test", description="test mappings"
    )
    assert toml_writer.dumps(toml_data) == snapshot
