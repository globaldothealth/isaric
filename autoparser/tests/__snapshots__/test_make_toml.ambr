# serializer version: 1
# name: test_make_toml
  '''
  [adtl]
  name = "test"
  description = "test mappings"
  
  [adtl.tables.subject]
  kind = "groupBy"
  groupBy = "subject_id"
  aggregation = "lastNotNull"
  schema = "../../schemas/dev/subject.schema.json"
  
  [adtl.tables.visit]
  kind = "groupBy"
  groupBy = "subject_id"
  aggregation = "lastNotNull"
  schema = "../../schemas/dev/visit.schema.json"
  
  [adtl.tables.observation]
  kind = "oneToMany"
  schema = "../../schemas/dev/observation.schema.json"
  common = { visit_id = { field = "subjid", sensitive = true } }
  
  [adtl.defs]
  "Y/N/NK" = { values = { 1 = true, 2 = false } }
  
  [adtl.defs."discharged alive/hospitalization/transfer to other facility/death/palliative discharge".values]
  1 = "discharged alive"
  2 = "hospitalization"
  3 = "transfer to other facility"
  4 = "death"
  5 = "palliative discharge"
  
  [subject.admission_date]
  combinedType = "firstNonNull"
  
  [[subject.admission_date.fields]]
  field = "ccm_b_icu_hodat"
  description = "ICU ADMISSION DATE"
  
  [[subject.admission_date.fields]]
  field = "hostdat_transfer"
  description = "2.5.2 If YES: Admission date at transfer facility"
  
  [[subject.admission_date.fields]]
  field = "date"
  description = "Date"
  
  [[subject.admission_date.fields]]
  field = "icu_hostdat"
  description = "Date of ICU admission:"
  
  [subject.age]
  field = "age_estimateyears"
  description = "Age"
  
  [subject.date_death]
  field = "f_datedeath"
  description = "5.5 Date Of Death"
  
  [subject.enrolment_date]
  field = "dsstdat"
  description = "<acronym title=\"Date of enrolment into the study or for in-patients is the date that COVID-19 was first assessed as suspected or confirmed by a clinician.\">Enrolment date / first COVID-19 assessment date:</acronym>"
  
  [subject.has_asthma]
  field = "asthma_mhyn"
  description = "<acronym title=\"Clinician-diagnosed asthma\">Asthma <i>(physician diagnosed)</i></acronym>"
  ref = "Y/N/NK"
  
  [subject.has_chronic_cardiac_disease]
  field = "chroniccard_mhyn"
  description = "<acronym title=\"Include any of coronary artery disease, heart failure, congenital heart disease, cardiomyopathy, rheumatic heart disease.\">Chronic cardiac disease <i>(not hypertension)</i></acronym>"
  ref = "Y/N/NK"
  
  [subject.has_chronic_hematologic_disease]
  field = "chronhaemo_mhyn"
  description = "<acronym title=\"Any long-term disorder of the red or white blood cells, platelets or coagulation system requiring regular or intermittent treatment.  Do not include leukaemia, lymphoma or myeloma, which should be entered under malignancy. Do not include iron-deficiency anaemia which is explained by diet or chronic blood loss.\">Chronic hematologic disease</acronym>"
  ref = "Y/N/NK"
  
  [subject.has_chronic_kidney_disease]
  field = "renal_mhyn"
  description = "<acronym title=\"Please include any of clinician-diagnosed chronic kidney disease, chronic estimated glomerular filtration rate < 60mL/min/1.73m2, history of kidney transplantation\">Chronic kidney disease</acronym>"
  ref = "Y/N/NK"
  
  [subject.has_liver_disease]
  combinedType = "any"
  
  [[subject.has_liver_disease.fields]]
  field = "mildliv_mhyn"
  description = "<acronym title=\"This is defined as cirrhosis without portal hypertension or chronic hepatitis\">Mild liver disease</acronym>"
  ref = "Y/N/NK"
  
  [[subject.has_liver_disease.fields]]
  field = "modliver_mhyn"
  description = "<acronym title=\"This is defined as cirrhosis with portal hypertension, with or without bleeding or a history of variceal bleeding\">Moderate or severe liver disease</acronym>"
  ref = "Y/N/NK"
  
  [visit.date_outcome]
  field = "dsstdtc"
  description = "<acronym title=\"Please state the date for the outcome listed above.\">Outcome date</acronym>"
  
  [visit.icu_admission]
  field = "icu_hostdat"
  description = "Date of ICU admission:"
  
  [visit.outcome]
  field = "dsterm"
  description = "<acronym title=\"Definition available in the CRF Completion Guide - see your FILE REPOSITORY for a copy.\">Outcome:</acronym>"
  values = { 1 = "discharged", 3 = "transferred", 4 = "death", 5 = "discharged" }
  
  [visit.treatment_anticoagulation]
  field = "anticoag_cmyn"
  description = "Anticoagulation?"
  ref = "Y/N/NK"
  
  [visit.treatment_antifungal_agent]
  field = "antifung_cmyn"
  description = "<acronym title=\"\"Antifungal agent\" refers to any agent(s) prescribed specifically to treat systemic or topical infections caused by fungi. Complete definition available in the CRF Completion Guide - see your FILE REPOSITORY for a copy.\">Antifungal agent?</acronym>"
  ref = "Y/N/NK"
  
  [visit.treatment_antivirals]
  field = "antiviral_cmyn"
  description = "<acronym title=\"Record all antivirals administered from date of admission or during the hospitalisation.  Record the total number of days the treatment was given. Complete definition available in the CRF Completion Guide - see your FILE REPOSITORY for a copy.\">Antiviral or COVID-19 targeted agent?</acronym>"
  ref = "Y/N/NK"
  
  [[observation]]
  name = "abdominal_pain"
  phase = "admission"
  date = { ref = "admissionDateHierarchy" }
  if = { abdopain_ceoccur_v2 = { "!=" = 3 } }
  is_present = { field = "abdopain_ceoccur_v2", description = "Abdominal pain", ref = "Y/N/NK" }
  
  [[observation]]
  name = "altered_consciousness_confusion"
  phase = "admission"
  date = { ref = "admissionDateHierarchy" }
  if = { confusion_ceoccur_v2 = { "!=" = 3 } }
  
  [observation.is_present]
  field = "confusion_ceoccur_v2"
  description = "Altered consciousness / confusion"
  ref = "Y/N/NK"
  
  [[observation]]
  name = "anorexia"
  phase = "admission"
  date = { ref = "admissionDateHierarchy" }
  if = { anorexia_ceoccur_v2 = { "!=" = 3 } }
  is_present = { field = "anorexia_ceoccur_v2", description = "Anorexia", ref = "Y/N/NK" }
  
  [[observation]]
  name = "glasgow_coma_score"
  phase = "study"
  date = { ref = "dailyDateHierarchy" }
  if = { daily_gcs_vsorres = { "!=" = "" } }
  
  [observation.value]
  field = "daily_gcs_vsorres"
  description = "<acronym title=\"Please state the lowest GCS recorded.  Complete definition available in the CRF Completion Guide - see your FILE REPOSITORY for a copy.\">Glasgow Coma Score (GCS / 15):</acronym>"
  
  [[observation]]
  name = "headache"
  phase = "admission"
  date = { ref = "admissionDateHierarchy" }
  if = { headache_ceoccur_v2 = { "!=" = 3 } }
  is_present = { field = "headache_ceoccur_v2", description = "Headache", ref = "Y/N/NK" }
  
  [[observation]]
  name = "heart_rate_bpm"
  phase = "admission"
  date = { ref = "admissionDateHierarchy" }
  if = { hr_vsorres = { "!=" = "" } }
  
  [observation.value]
  field = "hr_vsorres"
  description = "<acronym title=\"Enter the heart rate measured in beats per minute. This may be measured manually or by electronic monitoring.\">Heart Rate</acronym>"
  
  [[observation]]
  name = "heart_rate_bpm"
  phase = "study"
  date = { ref = "dailyDateHierarchy" }
  if = { daily_hr_vsorres = { "!=" = "" } }
  
  [observation.value]
  field = "daily_hr_vsorres"
  description = "<acronym title=\"Enter the heart rate measured in beats per minute. This may be measured manually or by electronic monitoring.\">Heart Rate</acronym>"
  
  '''
# ---
